### PR TITLE
Fix resizing crashes & potentially eject crashes

### PR DESCRIPTION
**Pull request by Godnoken**
This fixes crashes on resize/reset with DX9 games and likely most of the ejection crashes as discussed here https://github.com/veeenu/hudhook/issues/72

Funny enough - I had seen that Github issue before but didn't realize that was why I had issues with some DX9 games crashing on launch due to them calling the resize/reset function multiple times. Thanks to https://github.com/veeenu/hudhook/issues/73#issuecomment-1412476482 I understood where to look.

I have not yet tried ejecting but I assume that it will still crash sometimes as mentioned in the issue, unless this reset fix makes that go away completely.

### CODE REVIEW

File: src/hooks/dx9.rs [hunk 1]
```diff
@@ -123,8 +125,8 @@ unsafe extern "system" fn imgui_wnd_proc(
     WPARAM(wparam): WPARAM,
     LPARAM(lparam): LPARAM,
 ) -> LRESULT {
-    if IMGUI_RENDERER.is_some() {
-        match IMGUI_RENDERER.as_mut().unwrap().try_lock() {
+    if IMGUI_RENDERER.get().is_some() {
+        match IMGUI_RENDERER.get_mut().unwrap().try_lock() {
```
Comment (veeenu) [inline src/hooks/dx9.rs:129]: Small nit, just noticed this now; even though it's not pertinent to this PR, while we're at it I'd lighten up the syntax by compressing the if and match in the same statement:
```suggestion
        if let Some(Ok(renderer)) = IMGUI_RENDERER.get_mut().map(|m| m.try_lock()) {
```

**Review by veeenu — CHANGES_REQUESTED** [review]
Thank you for your PR, looks pretty neat!

Can you try ejecting and run a few more manual tests? Once that's done, I'd say lgtm.

File: src/hooks/dx9.rs (hunk 1 above)
Comment (Godnoken) [reply src/hooks/dx9.rs]: Done. However, I had to use Some(Some()) instead, a bit ugly but eh, it was ugly anyways 😄

**Comment by Godnoken** [comment]
> Thank you for your PR, looks pretty neat!

Cheers! My first real contribution to OSS.

> Can you try ejecting and run a few more manual tests? Once that's done, I'd say lgtm.

Done & done.

I've had no issues with resize crashes so far. Ejecting does unfortunately still crash sometimes. Some games work flawlessly while other doesn't seem to like it all the time.

I tried;

- M&B: Warband
- W40K
- Shelter 2
- Fallout 3
- Killing Floor
- Tropico 4
- Victoria II
-

File: src/hooks/dx9.rs (hunk 1 above)
Comment (veeenu) [reply src/hooks/dx9.rs]: Oh, in that case, my bad, thought it was an `Option<Result>`. In those cases, `and_then` in place of `map` has you covered 😅

**Comment by veeenu** [comment]
> > Thank you for your PR, looks pretty neat!
> 
> Cheers! My first real contribution to OSS.

💪 

> I've had no issues with resize crashes so far. Ejecting does unfortunately still crash sometimes. Some games work flawlessly while other doesn't seem to like it all the time.

No worries, I think it's heavily dependent on what the game does anyway, and besides even DX11&12 are affected by the same instability.
 
> Do you think it is a possibility that SetWindowLongA/PtrA is calling it on the wrong window in some games? DX11 and DX12 uses the swapchain to get the correct window, while DX9 uses [D3DDEVICE_CREATION_PARAMETERS](https://learn.microsoft.com/en-us/windows/win32/direct3d9/d3ddevice-creation-parameters). I'm thinking that hFocusWindow isn't guaranteed to be the correct window if the game uses multiple.

I think most of our methods are heuristics anyway -- if we find a better one than `hFocusWindow`, then I'm all for switching over. It would probably be worth even if it was a more expensive method, as it's going to be run only once anyway (or once per resize, at worst).

> I just noticed the OpenGL3 hook doesn't have a proper cleanup fn either, I'll get that fixed too.

Great, thank you 🙏

**Comment by Godnoken** [comment]
> No worries, I think it's heavily dependent on what the game does anyway, and besides even DX11&12 are affected by the same instability.

Oh really? Hmm.. 

> I think most of our methods are heuristics anyway -- if we find a better one than `hFocusWindow`, then I'm all for switching over. It would probably be worth even if it was a more expensive method, as it's going to be run only once anyway (or once per resize, at worst).

I'm trying it out on OpenGL3 now & it also crashes sometimes. The OpenGL implementation uses [WindowFromDC](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-windowfromdc), so in that case - that's a third option that doesn't work.

I'll see if I can catch the crash error & investigate further.

Btw - [SetWindowLongPtr](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowlongptra) is the only one needed for both 32-bit and 64-bit according to its description. Shall I drop SetWindowLong from all hooks?

File: src/hooks/dx9.rs (hunk 1 above)
Comment (Godnoken) [reply src/hooks/dx9.rs]: I'm not too sure of how to write it as my idiomatic Rust skills are not quite there yet.. 😄 

I'll leave it as is for now.

**Comment by Godnoken** [comment]
I've never used debuggers in games like this so I honestly am not too sure of how to use it.. `User-mode data execution prevention (DEP) violation at location` is the error I got from lldb when crashing. Doesn't tell me much. Eitherway, DX9 seems to crash either before or during `FreeLibraryAndExitThread` while OpenGL seemingly only crashes during `FreeLibraryAndExitThread`.

I feel like this is out of my depth and I have more pressing things to take care of at the moment.

Think we're good to merge this for now?

File: src/hooks/opengl3.rs [hunk 2]
```diff
@@ -184,7 +184,13 @@ impl ImguiRenderer {
         self.renderer.render(&mut self.ctx);
     }
 
-    unsafe fn cleanup(&mut self) {}
+    unsafe fn cleanup(&mut self) {
+        #[cfg(any(target_arch = "aarch64", target_arch = "x86_64"))]
+        dbg!(SetWindowLongPtrA(self.game_hwnd, GWLP_WNDPROC, self.wnd_proc as usize as isize));
+
+        #[cfg(target_arch = "x86")]
+        dbg!(SetWindowLongA(self.game_hwnd, GWLP_WNDPROC, self.wnd_proc as usize as i32));
```
Comment (veeenu) [inline src/hooks/opengl3.rs:192]: Could you please remove the `dbg!` here?

I think we _should_ probably manage the resulting errors but we should emit a well formed log entry instead. We can do it in the future, in a dedicated refactoring PR.

File: src/hooks/dx9.rs [hunk 3]
```diff
@@ -123,21 +125,15 @@ unsafe extern "system" fn imgui_wnd_proc(
...
-                LPARAM(lparam),
-                imgui_renderer,
-                IMGUI_RENDER_LOOP.get().unwrap(),
-            ),
-            None => {
-                debug!("Could not lock in WndProc");
-                DefWindowProcW(hwnd, umsg, WPARAM(wparam), LPARAM(lparam))
-            },
-        }
+    if let Some(Some(imgui_renderer)) = IMGUI_RENDERER.get_mut().map(|m| m.try_lock()) {
```
Comment (veeenu) [inline src/hooks/dx9.rs:128]:
```suggestion
    if let Some(imgui_renderer) = IMGUI_RENDERER.get_mut().and_then(|m| m.try_lock()) {
```

**Review by veeenu — CHANGES_REQUESTED** [review]
Couple minor things and then we're good to go.

**Comment by veeenu** [comment]
> Scratch that, 32-bit implementation of SetWindowLongPtr doesn't exist in the windows crate until 0.41.0 and we're one 0.39.0. Any plans to bump up?

Absolutely. Haven't been able to keep tabs on `windows-rs` but I'll try and make the bump PR as soon as I get the chance.

> I feel like this is out of my depth and I have more pressing things to take care of at the moment.
> 
> Think we're good to merge this for now?

Regardless of that, I feel like this PR is very worthwhile as the semantics around `OnceCell` are better than `Option`.
So, resize crashes are gone and eject crashes are occasional still? If that's the case, it's still a huge improvement.

File: src/hooks/opengl3.rs (hunk 2 above)
Comment (Godnoken) [reply src/hooks/opengl3.rs:192]: Woops! Didn't mean to leave it like that, my bad 😆

**Comment by Godnoken** [comment]
> Absolutely. Haven't been able to keep tabs on `windows-rs` but I'll try and make the bump PR as soon as I get the chance.

Excellent!

> Regardless of that, I feel like this PR is very worthwhile as the semantics around `OnceCell` are better than `Option`. So, resize crashes are gone and eject crashes are occasional still? If that's the case, it's still a huge improvement.

I have personally not experienced any resize crashes but eject crashes do indeed still happen sometimes.

There's probably a better way to do the resizing too as it is not as smooth as in the other renderers - but it works, and how often do you resize really?

**Comment by veeenu** [comment]
> There's probably a better way to do the resizing too as it is not as smooth as in the other renderers - but it works, and how often do you resize really?

It is something that I expect regular users only do when they want to switch their game resolution up or down every once in a while; I see most games nowadays only support a few different somewhat standard resolutions and handle all of that themselves anyway, so yeah, not terribly critical. It's just important that it works correctly but it doesn't need to be mega fast.

**Review by veeenu — APPROVED** [review]
LGTM!

**Pull request merged**